haystack/preview/components/samplers/top_p.py

```diff
@@ -0,0 +1,144 @@
+
+
+logger = logging.getLogger(__name__)
+
+
+with LazyImport(
+    message="Run 'pip install transformers[torch,sentencepiece]==4.32.1 sentence-transformers>=2.2.0'"
+) as torch_and_transformers_import:
+    import torch
+    from sentence_transformers import CrossEncoder
+    from haystack.modeling.utils import initialize_device_settings  # pylint: disable=ungrouped-imports
+
+
+@component
+class TopPSampler:
+    """
+    Selects documents based on the cumulative probability of the similarity scores between the
+    query and the documents using top p sampling.
+
+    Top p sampling selects a subset of the most relevant data points from a larger set of data. The technique
+    involves calculating the cumulative probability of the scores of each data point, and then
+    selecting the top p percent of data points with the highest cumulative probability.
+
+    In the context of TopPSampler, the `run()` method takes in a query and a set of documents,
+    calculates the similarity scores between the query and the documents, and then filters
+    the documents based on the cumulative probability of these scores. The TopPSampler provides a
+    way to efficiently select the most relevant documents based on their similarity to a given query.
+
+    Usage example:
+    ```
+    from haystack.preview import Document
+    from haystack.preview.components.samplers import TopPSampler
+
+    sampler = TopPSampler(top_p=0.95)
+    docs = [Document(text="Paris"), Document(text="Berlin")]
+    query = "City in Germany"
+    output = sampler.run(query=query, documents=docs)
+    docs = output["documents"]
+    assert len(docs) == 1
+    assert docs[0].text == "Berlin"
+    ```
+    """
+
+    def __init__(
+        self,
+        model_name_or_path: Union[str, Path] = "cross-encoder/ms-marco-MiniLM-L-6-v2",
+        top_p: Optional[float] = 1.0,
+        score_field: Optional[str] = "score",
+        devices: Optional[List[Union[str, "torch.device"]]] = None,
+    ):
+        """
+
+        :param model_name_or_path: Path to a pretrained sentence-transformers model.
+        :param top_p: Cumulative probability threshold for filtering the documents (usually between 0.9 and 0.99).
+        `False` ensures at least one document is returned. If `strict` is set to `True`, then no documents are returned.
+        :param score_field: The name of the field that should be used to store the scores a document's meta data.
+        :param devices: List of torch devices (for example, cuda:0, cpu, mps) to limit inference to specific devices.
+        """
+        torch_and_transformers_import.check()
+        super().__init__()
+
+        self.model_name_or_path = model_name_or_path
+        self.top_p = top_p
+        self.score_field = score_field
+        self.devices, _ = initialize_device_settings(devices=devices)
+        self.cross_encoder = CrossEncoder(model_name_or_path, device=str(self.devices[0]))
```

Comment:
Oh something to consider. I wonder if we should make this node not contain any models? Instead it could come after any node that produces similarity scores that we could then top-p sample. For example, then we could add this node right after an EmbeddingRetriever, a CohereRanker, SentenceTransformersRanker, etc. I think that might increase the utility of this instead of requiring a SentenceTransformers Cross Encoder model to be used.

Reply:
Oooh very good idea. Let's make similarity scoring optional in this node; that's easiest, no?

Reply:
We can perhaps look for similarity scores in two "places". The first one is an array with similarity scores passed as a run method parameter, and then the second - meta["score"] field of each document. If none are present, then we calculate similarity scores with a cross encoder. LMK

Reply:
I really like the idea of splitting rankers (that compute scores) and samplers (that filter documents), but I have a doubt too: is there any real usecase where we don't need to recompute the scores? Initially I thought it was great but now I can hardly find a usecase where this distinction would be useful. Do you have any example?

Reply:
One could use CohereRanker (it produces scores and pushes them to the metadata of the document) and then after it position TopPSampler, which will then sample based on those scores. Here (in TopPSampler) we don't need to recompute scores because CohereRanker (which the user trusts and wants) already computed them. That's what @sjrl meant here.

Reply:
> That's what @sjrl meant here.

Yup that is exactly right. Also, we can directly use the scores provided by an Embedding Retriever. We don't have to use Cross-Encoder scores.

Reply:
> We can perhaps look for similarity scores in two "places". The first one is an array with similarity scores passed as a run method parameter, and then the second - meta["score"] field of each document. If none are present, then we calculate similarity scores with a cross encoder. 

Yeah, I'm mostly happy with this idea. The only thing I'd like to address and am unsure about is if the similarity scores are provided I would prefer not still having to load the cross-encoder model. So basically make the cross-encoder model optional and only load it if it is specified by the user. Otherwise, we risk wasting resources by loading a model we don't use.

Reply:
Ok, good point, will do that now!

Reply:
Hey I'm sorry to chime in again late, but I'm not super happy about a component that may or may not compute scores depending on how is it initialized. I'd rather have one component that does that (a Ranker) and another that does not (a Sampler). Do you see some drawbacks with this approach?

It's not an abstract issue implementation-wise: a sampler should not need to import torch and transformers if it's not computing the scores, just to name one. And I see the implementation grew, which usually is a sign that this component is doing many things at once.

In this way we would also have a Ranker immediately and we can make the e2e test @sjrl was mentioning below.